src/client/UrqlClientTypes.re

```diff
@@ -0,0 +1,143 @@
+type t;
+
+/* The result of executing a GraphQL request.
+   Consists of optional data and errors fields. */
+type executionResult = {
+  errors: option(array(UrqlCombinedError.graphQLError)),
+  data: option(Js.Json.t),
+  extensions: Js.Json.t,
+};
+
+/* OperationType for the active operation.
+   Use with operationTypeToJs for proper conversion to strings. */
+[@bs.deriving jsConverter]
+type operationType = [
+  | [@bs.as "query"] `Query
+  | [@bs.as "mutation"] `Mutation
+  | [@bs.as "subscription"] `Subscription
+  | [@bs.as "teardown"] `Teardown
+];
+
+/* Cache outcomes for operations. */
+[@bs.deriving jsConverter]
+type cacheOutcome = [
+  | [@bs.as "miss"] `Miss
+  | [@bs.as "partial"] `Partial
+  | [@bs.as "hit"] `Hit
+];
+
+/* Debug information on operations. */
+type operationDebugMeta = {
+  source: option(string),
+  cacheOutcome: option(cacheOutcome),
+  networkLatency: option(int),
+  startTime: option(int),
+};
+
+/* The operation context object for a request. */
+type operationContext = {
+  fetchOptions: option(Fetch.requestInit),
+  requestPolicy: UrqlTypes.requestPolicy,
+  url: string,
+  meta: option(operationDebugMeta),
+  pollInterval: option(int),
+};
+
+/* A partial operation context, which can be passed as the second
+   optional argument to executeQuery, executeMutation, and executeSubscription.
+   Simply makes all properties on operationContext optional. */
+[@bs.deriving {abstract: light}]
+type partialOperationContext = {
+  [@bs.optional]
+  fetchOptions: Fetch.requestInit,
+  [@bs.optional]
+  requestPolicy: UrqlTypes.requestPolicy,
+  [@bs.optional]
+  url: string,
+  [@bs.optional]
+  debugMeta: operationDebugMeta,
+  [@bs.optional]
+  pollInterval: int,
+};
+
+module PartialOperationContextJs = {
+  [@bs.deriving {abstract: light}]
```

Comment:
Sadly had to keep around this `[@bs.deriving abstract]`. This is because we can't have partial objects with records; all fields on the record need to be explicitly specified. This meant that we'd need to create records for `partialOperationContext` with `None` for all fields that the user didn't want to modify, which then explicitly sets these to `undefined`. When `urql` receives these it'll overwrite the defaults with `undefined`, which blows up operations 😭 

`abstract: light` just helps us get tighter names for the getters and setters generated by BuckleScript.